numpy/core/_methods.py

```diff
@@ -56,14 +56,28 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
 def _all(a, axis=None, dtype=None, out=None, keepdims=False):
     return umr_all(a, axis, dtype, out, keepdims)
 
-def _count_reduce_items(arr, axis):
-    if axis is None:
-        axis = tuple(range(arr.ndim))
-    if not isinstance(axis, tuple):
-        axis = (axis,)
-    items = 1
-    for ax in axis:
-        items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
+def _count_reduce_items(arr, axis, keepdims=False, where=True):
+    if isinstance(where, bool):
```

Comment:
```suggestion
    # fast-path for the default case
    if where is True:
```

Reply:
Good suggestion, I will included it in the change.
However what should the expected result of `where=None` and `where=False` be? Right now, I do not catch those cases separately, so they go through the broadcast branch. This could be more efficient with something like this:
```
if where is None or where is False:
    # no elements in slice
    return nt.intp(0
```
but of course if this is not _legal_ case, it might not be needed.

If you run `arr.sum(where=False)` or `arr.sum(where=None)`, the result is always 0. However due to the division in `mean`, it turns out to be `nan` with a warning. Should this be caught and zero returned or is the warning alright since `False` and `None` are not among the expected inputs to `where`?

Reply:
> Right now, I do not catch those cases separately, so they go through the broadcast branch.

The best way to make sure your corner cases are readonable is to not have to special case them at all.

> This could be more efficient with 

Remember, the reason we care about optimizing `where=True` is because it's the default, and the path taken by all existing code using `mean` today. We don't have any reason to care about optimizing other corner cases, all that matters is that we get them correct.

Reply:
Also, `where=None` is only legal for the same reason as `where=""` or `where=MyFalsyObject()`.

Reply:
> The best way to make sure your corner cases are reasonable is to not have to special case them at all. Sending them through the broadcast branch is the right thing to do (even if your optimization was correct).
> Remember, the reason we care about optimizing `where=True` is because it's the default, and the path taken by all existing code using `mean` today. We don't have any reason to care about optimizing other corner cases, all that matters is that we get them correct.

OK makes sense, thank you for clarifying!